staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/interfaces.go

```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	"context"
+
+	"k8s.io/apiserver/pkg/admission"
+)
+
+// Source can list dynamic policy plugins.
+type Source[Hook any] interface {
+	Hooks() []Hook
+	Run(ctx context.Context) error
+	HasSynced() bool
+}
+
+// Dispatcher dispatches policy call to a list of policies with admission attributes as argument.
+type Dispatcher[Hook any] interface {
+	// Dispatch a request to the policies. Dispatcher may choose not to
+	// call a hook, either because the rules of the hook does not match, or
+	// the namespaceSelector or the objectSelector of the hook does not
```

Comment:
Given this explanation, should Hook have a type constraint that provides match, namespaceSelector and objectSelector access?

Reply:
That would be an interesting idea that works for webhooks, but for policy we hooks both the Policy and Binding have matching logic and selectors that needs to be handled specially by the dispatcher that I feel would be complicated by conforming it to such an interface.

Reply:
An alternative might be to use a minimal (or even empty) interface that documents the expectations of a Hook.

Reply:
Added